### DISCUSSION TITLE
Promblem when building PETSc

### DISCUSSION THREAD

**Discussion by zsh151** [neams-th-coe/cardinal#445] [Q&A] [answered]
I Build PETSc in the conda environment of moose. When I ran ./contrib/moose/scripts/update_and_rebuild_petsc.sh , the following error appeared.
<img width="377" alt="image" src="https://user-images.githubusercontent.com/110072634/182544982-f041d094-df6d-42b3-ac42-3e1502bd5cc8.png">

**Comment by aprilnovak** [comment, marked as answer]
I'm a bit confused on your wording - are you building PETSc using the scripts, or are you intending to use the PETSc that ships with MOOSE's conda environment? If you are doing the latter, you don't need to run the `update_and_rebuild...` scripts.

But in any case, if something goes wrong at the PETSc/libMesh stage when building Cardinal, 95% of the time it is a problem that you would also get when trying to build MOOSE as a standalone project. In other words, my guess is that this problem you're seeing isn't specific to Cardinal. For this reason, the [MOOSE discussions page](https://github.com/idaholab/moose/discussions)/Slack would be the best place to get help with this issue. I will move this question over there to try to get help.

**Reply by zsh151** [reply]
I  intended to use the PETSc that ships with MOOSE's conda environment, but when I make, another error appeared. Then I tried to rebuild the PETSc. Finally, it turns out the error has nothing to do with the PETSc. I rebuild MOOSE and now that error doesn't appear,

But a new error has occurred when I run make. I have updated MOOSE but it still can't be solved.

<img width="363" alt="image" src="https://user-images.githubusercontent.com/110072634/182744571-fb18941a-3331-4bfa-96fc-2c62a4ae714e.png">

**Reply by aprilnovak** [reply]
Could you please post the entire screen output that you see? The error is happening during the Nek make stage, but the message is going to appear earlier than what's viewable in your screenshot.

You can pipe the output of make to a file like this:


```
make -j4 > file.txt
```

**Reply by zsh151** [reply]
Here it is.
[file.txt.txt](https://github.com/neams-th-coe/cardinal/files/9256236/file.txt.txt)

**Reply by zsh151** [reply]
I'm also little confused that should I build OpenMC and nekRS first in the MOOSE's conda environment?

**Reply by aprilnovak** [reply]
Cardinal automatically builds OpenMC and NekRS for you - no need to do anything separately. The errors I'm seeing are regarding HYPRE, and look like this:

```
...
libnekrs.so: undefined reference to `HYPRE_BoomerAMGCreate'
libnekrs.so: undefined reference to `HYPRE_BoomerAMGSetTol'
```

Can you try explicitly pointing to the HYPRE in MOOSE's conda environment, and then re-build:

```
export HYPRE_DIR=$CONDA_PREFIX
cd cardinal
rm -rf build/ install/
make -j8
```

**Reply by aprilnovak** [reply]
@zsh151 , was the above able to work for you?

**Reply by zsh151** [reply]
It worked, but still exist other errors

**Reply by aprilnovak** [reply]
If you post your errors here I can take a look?

**Reply by zsh151** [reply]
Sure, please wait a minute.
I found my question is similar with [https://github.com/neams-th-coe/cardinal/discussions/362](url) which is not solved yet.

**Reply by zsh151** [reply]
I have set `export HYPRE_DIR=$CONDA_PREFIX`, but it seems the error still exist.

[make.txt](https://github.com/neams-th-coe/cardinal/files/9264438/make.txt)

```
[100%] Linking CXX executable nekrs
/home/en/mambaforge3/envs/moose/bin/../lib/gcc/x86_64-conda-linux-gnu/10.4.0/../../../../x86_64-conda-linux-gnu/bin/ld: warning: libHYPRE-2.23.0.so, needed by libnekrs.so, not found (try using -rpath or -rpath-link)
/home/en/mambaforge3/envs/moose/bin/../lib/gcc/x86_64-conda-linux-gnu/10.4.0/../../../../x86_64-conda-linux-gnu/bin/ld: warning: libmpifort.so.12, needed by libnekrs.so, not found (try using -rpath or -rpath-link)
/home/en/mambaforge3/envs/moose/bin/../lib/gcc/x86_64-conda-linux-gnu/10.4.0/../../../../x86_64-conda-linux-gnu/bin/ld: warning: libgfortran.so.5, needed by libnekrs.so, not found (try using -rpath or -rpath-link)
/home/en/mambaforge3/envs/moose/bin/../lib/gcc/x86_64-conda-linux-gnu/10.4.0/../../../../x86_64-conda-linux-gnu/bin/ld: warning: libquadmath.so.0, needed by libnekrs.so, not found (try using -rpath or -rpath-link)
/home/en/mambaforge3/envs/moose/bin/../lib/gcc/x86_64-conda-linux-gnu/10.4.0/../../../../x86_64-conda-linux-gnu/bin/ld: libnekrs.so: undefined reference to `HYPRE_IJVectorGetObject'
/home/en/mambaforge3/envs/moose/bin/../lib/gcc/x86_64-conda-linux-gnu/10.4.0/../../../../x86_64-conda-linux-gnu/bin/ld: libnekrs.so: undefined reference to `_gfortran_st_write@GFORTRAN_8'
/home/en/mambaforge3/envs/moose/bin/../lib/gcc/x86_64-conda-linux-gnu/10.4.0/../../../../x86_64-conda-linux-gnu/bin/ld: libnekrs.so: undefined reference to `HYPRE_IJMatrixCreate'
/home/en/mambaforge3/envs/moose/bin/../lib/gcc/x86_64-conda-linux-gnu/10.4.0/../../../../x86_64-conda-linux-gnu/bin/ld: libnekrs.so: undefined reference to `HYPRE_IJVectorInitialize'
/home/en/mambaforge3/envs/moose/bin/../lib/gcc/x86_64-conda-linux-gnu/10.4.0/../../../../x86_64-conda-linux-gnu/bin/ld: libnekrs.so: undefined reference to `HYPRE_BoomerAMGSetCoarsenType'
/home/en/mambaforge3/envs/moose/bin/../lib/gcc/x86_64-conda-linux-gnu/10.4.0/../../../../x86_64-conda-linux-gnu/bin/ld: libnekrs.so: undefined reference to `HYPRE_BoomerAMGSetLevelNonGalerkinTol'
/home/en/mambaforge3/envs/moose/bin/../lib/gcc/x86_64-conda-linux-gnu/10.4.0/../../../../x86_64-conda-linux-gnu/bin/ld: libnekrs.so: undefined reference to `HYPRE_IJVectorCreate'
/home/en/mambaforge3/envs/moose/bin/../lib/gcc/x86_64-conda-linux-gnu/10.4.0/../../../../x86_64-conda-linux-gnu/bin/ld: libnekrs.so: undefined reference to `HYPRE_BoomerAMGSetNonGalerkinTol'
/home/en/mambaforge3/envs/moose/bin/../lib/gcc/x86_64-conda-linux-gnu/10.4.0/../../../../x86_64-conda-linux-gnu/bin/ld: libnekrs.so: undefined reference to `HYPRE_BoomerAMGSetMaxIter'
/home/en/mambaforge3/envs/moose/bin/../lib/gcc/x86_64-conda-linux-gnu/10.4.0/../../../../x86_64-conda-linux-gnu/bin/ld: libnekrs.so: undefined reference to `HYPRE_IJMatrixAddToValues'
/home/en/mambaforge3/envs/moose/bin/../lib/gcc/x86_64-conda-linux-gnu/10.4.0/../../../../x86_64-conda-linux-gnu/bin/ld: libnekrs.so: undefined reference to `_gfortran_transfer_character_write@GFORTRAN_8'
/home/en/mambaforge3/envs/moose/bin/../lib/gcc/x86_64-conda-linux-gnu/10.4.0/../../../../x86_64-conda-linux-gnu/bin/ld: libnekrs.so: undefined reference to `HYPRE_IJVectorAssemble'
/home/en/mambaforge3/envs/moose/bin/../lib/gcc/x86_64-conda-linux-gnu/10.4.0/../../../../x86_64-conda-linux-gnu/bin/ld: libnekrs.so: undefined reference to `HYPRE_IJMatrixDestroy'
/home/en/mambaforge3/envs/moose/bin/../lib/gcc/x86_64-conda-linux-gnu/10.4.0/../../../../x86_64-conda-linux-gnu/bin/ld: libnekrs.so: undefined reference to `HYPRE_IJMatrixAssemble'
/home/en/mambaforge3/envs/moose/bin/../lib/gcc/x86_64-conda-linux-gnu/10.4.0/../../../../x86_64-conda-linux-gnu/bin/ld: libnekrs.so: undefined reference to `HYPRE_BoomerAMGSetAggNumLevels'
/home/en/mambaforge3/envs/moose/bin/../lib/gcc/x86_64-conda-linux-gnu/10.4.0/../../../../x86_64-conda-linux-gnu/bin/ld: libnekrs.so: undefined reference to `HYPRE_BoomerAMGSetup'
/home/en/mambaforge3/envs/moose/bin/../lib/gcc/x86_64-conda-linux-gnu/10.4.0/../../../../x86_64-conda-linux-gnu/bin/ld: libnekrs.so: undefined reference to `HYPRE_BoomerAMGSetPrintLevel'
/home/en/mambaforge3/envs/moose/bin/../lib/gcc/x86_64-conda-linux-gnu/10.4.0/../../../../x86_64-conda-linux-gnu/bin/ld: libnekrs.so: undefined reference to `_gfortran_st_write_done@GFORTRAN_8'
/home/en/mambaforge3/envs/moose/bin/../lib/gcc/x86_64-conda-linux-gnu/10.4.0/../../../../x86_64-conda-linux-gnu/bin/ld: libnekrs.so: undefined reference to `HYPRE_BoomerAMGSetMinCoarseSize'
/home/en/mambaforge3/envs/moose/bin/../lib/gcc/x86_64-conda-linux-gnu/10.4.0/../../../../x86_64-conda-linux-gnu/bin/ld: libnekrs.so: undefined reference to `HYPRE_IJMatrixGetValues'
/home/en/mambaforge3/envs/moose/bin/../lib/gcc/x86_64-conda-linux-gnu/10.4.0/../../../../x86_64-conda-linux-gnu/bin/ld: libnekrs.so: undefined reference to `HYPRE_BoomerAMGDestroy'
/home/en/mambaforge3/envs/moose/bin/../lib/gcc/x86_64-conda-linux-gnu/10.4.0/../../../../x86_64-conda-linux-gnu/bin/ld: libnekrs.so: undefined reference to `HYPRE_IJMatrixSetObjectType'
/home/en/mambaforge3/envs/moose/bin/../lib/gcc/x86_64-conda-linux-gnu/10.4.0/../../../../x86_64-conda-linux-gnu/bin/ld: libnekrs.so: undefined reference to `HYPRE_IJVectorDestroy'
/home/en/mambaforge3/envs/moose/bin/../lib/gcc/x86_64-conda-linux-gnu/10.4.0/../../../../x86_64-conda-linux-gnu/bin/ld: libnekrs.so: undefined reference to `HYPRE_IJMatrixSetValues'
/home/en/mambaforge3/envs/moose/bin/../lib/gcc/x86_64-conda-linux-gnu/10.4.0/../../../../x86_64-conda-linux-gnu/bin/ld: libnekrs.so: undefined reference to `HYPRE_BoomerAMGSolve'
/home/en/mambaforge3/envs/moose/bin/../lib/gcc/x86_64-conda-linux-gnu/10.4.0/../../../../x86_64-conda-linux-gnu/bin/ld: libnekrs.so: undefined reference to `_gfortran_stop_string@GFORTRAN_8'
/home/en/mambaforge3/envs/moose/bin/../lib/gcc/x86_64-conda-linux-gnu/10.4.0/../../../../x86_64-conda-linux-gnu/bin/ld: libnekrs.so: undefined reference to `HYPRE_IJMatrixInitialize'
/home/en/mambaforge3/envs/moose/bin/../lib/gcc/x86_64-conda-linux-gnu/10.4.0/../../../../x86_64-conda-linux-gnu/bin/ld: libnekrs.so: undefined reference to `_gfortran_transfer_integer_write@GFORTRAN_8'
/home/en/mambaforge3/envs/moose/bin/../lib/gcc/x86_64-conda-linux-gnu/10.4.0/../../../../x86_64-conda-linux-gnu/bin/ld: libnekrs.so: undefined reference to `HYPRE_IJVectorSetObjectType'
/home/en/mambaforge3/envs/moose/bin/../lib/gcc/x86_64-conda-linux-gnu/10.4.0/../../../../x86_64-conda-linux-gnu/bin/ld: libnekrs.so: undefined reference to `HYPRE_IJMatrixGetRowCounts'
/home/en/mambaforge3/envs/moose/bin/../lib/gcc/x86_64-conda-linux-gnu/10.4.0/../../../../x86_64-conda-linux-gnu/bin/ld: libnekrs.so: undefined reference to `HYPRE_BoomerAMGSetCycleNumSweeps'
/home/en/mambaforge3/envs/moose/bin/../lib/gcc/x86_64-conda-linux-gnu/10.4.0/../../../../x86_64-conda-linux-gnu/bin/ld: libnekrs.so: undefined reference to `HYPRE_IJMatrixGetObject'
/home/en/mambaforge3/envs/moose/bin/../lib/gcc/x86_64-conda-linux-gnu/10.4.0/../../../../x86_64-conda-linux-gnu/bin/ld: libnekrs.so: undefined reference to `HYPRE_IJVectorGetValues'
/home/en/mambaforge3/envs/moose/bin/../lib/gcc/x86_64-conda-linux-gnu/10.4.0/../../../../x86_64-conda-linux-gnu/bin/ld: libnekrs.so: undefined reference to `HYPRE_BoomerAMGSetCycleRelaxType'
/home/en/mambaforge3/envs/moose/bin/../lib/gcc/x86_64-conda-linux-gnu/10.4.0/../../../../x86_64-conda-linux-gnu/bin/ld: libnekrs.so: undefined reference to `HYPRE_IJVectorSetValues'
/home/en/mambaforge3/envs/moose/bin/../lib/gcc/x86_64-conda-linux-gnu/10.4.0/../../../../x86_64-conda-linux-gnu/bin/ld: libnekrs.so: undefined reference to `HYPRE_BoomerAMGSetInterpType'
/home/en/mambaforge3/envs/moose/bin/../lib/gcc/x86_64-conda-linux-gnu/10.4.0/../../../../x86_64-conda-linux-gnu/bin/ld: libnekrs.so: undefined reference to `HYPRE_BoomerAMGCreate'
/home/en/mambaforge3/envs/moose/bin/../lib/gcc/x86_64-conda-linux-gnu/10.4.0/../../../../x86_64-conda-linux-gnu/bin/ld: libnekrs.so: undefined reference to `HYPRE_BoomerAMGSetStrongThreshold'
/home/en/mambaforge3/envs/moose/bin/../lib/gcc/x86_64-conda-linux-gnu/10.4.0/../../../../x86_64-conda-linux-gnu/bin/ld: libnekrs.so: undefined reference to `_gfortran_concat_string@GFORTRAN_8'
/home/en/mambaforge3/envs/moose/bin/../lib/gcc/x86_64-conda-linux-gnu/10.4.0/../../../../x86_64-conda-linux-gnu/bin/ld: libnekrs.so: undefined reference to `HYPRE_BoomerAMGSetTol'
collect2: error: ld returned 1 exit status
make[3]: *** [CMakeFiles/nekrs-bin.dir/build.make:99: nekrs] Error 1
make[3]: Leaving directory '/home/en/cardinal/build/nekrs'
make[2]: *** [CMakeFiles/Makefile2:205: CMakeFiles/nekrs-bin.dir/all] Error 2
make[2]: Leaving directory '/home/en/cardinal/build/nekrs'
make[1]: *** [Makefile:136: all] Error 2
make[1]: Leaving directory '/home/en/cardinal/build/nekrs'
make: *** [/home/en/cardinal/config/nekrs.mk:45: build_nekrs] Error 2
make: *** Waiting for unfinished jobs....
```

**Reply by aprilnovak** [reply]
Can you temporarily try the alternate approach of building PETSc and HYPRE from the scripts?

```
unset HYPRE_DIR
cd cardinal
./contrib/moose/scripts/update_and_rebuild_petsc.sh
./contrib/moose/scripts/update_and_rebuild_libmesh.sh
rm -rf build/ install/
make -j8
```

I am wondering if there is some incompatibility with the conda environment that we missed. I'm sorry this isn't working as smoothly as possible!

**Reply by zsh151** [reply]
It is for these reasons that I initially tried to build petsc and libmesh under moose's conda environment. But the result is exactly what I had initially encountered. Maybe I should try to build Cardinal directly in base environment.
Anyway，I still appreciate your patience in answering my questions.

**Reply by milljm** [reply]
If you have installed the `moose-libmesh` package (which in turn installs the `moose-petsc` package as a dependency), then there is no need to build PETSc (or libMesh). If you try to  do so using `moose/scripts/update_and_rebuild_petsc.sh|libmesh.sh` you will run into errors or headaches later. These two Conda packages separately export environment variables which will conflict with your application in the end (if you try to provide it multiple locations of PETSc/libMesh):

```bash
(base) [rocky-8][~]> mamba create -n testing moose-libmesh
(base) [rocky-8][~]> mamba activate testing
(testing) [rocky-8][~]> printf "$PETSC_DIR\n$LIBMESH_DIR\n"
/data/contribs/conda3/envs/testing
/data/contribs/conda3/envs/testing/libmesh
```
In the above example, I now have PETSC_DIR and LIBMESH_DIR pointing at

- `/data/contribs/conda3/envs/testing`

- `/data/contribs/conda3/envs/testing/libmesh`

Not good if I also plan on using  _my own_ PETSc and libMesh.

If you wish to use Conda to obtain only the minimum so you may build your own PETSc and/or libMesh, then I would recommend installing only the `moose-mpich` Conda package (or maybe nothing at all, and just use your systems compilers).

With all that said 😄  We have been experiencing issues with Ubuntu 22.04 and building PETSc using the `update_and_rebuild_petsc.sh` script. What version of Ubuntu are you running by chance?

**Reply by zsh151** [reply]
Yes, I also think so. 

I have tried to build PETSc manually with only the necessary compilers instead of using moose' conda environment, but the same error persists.

[image](https://user-images.githubusercontent.com/110072634/183094094-cacc6f38-8447-4c4a-89ce-612a1305cba9.png)

I'm running the Ubuntu 18.04 now.

**Reply by milljm** [reply]
Can you attach the configure.log? Might contain some useful information on why `ar` is failing.

Also, you may get better results using modern compilers by installing `moose-mpich` from Conda.

**Reply by zsh151** [reply]
> Can you temporarily try the alternate approach of building PETSc and HYPRE from the scripts?
> 
> ```
> unset HYPRE_DIR
> cd cardinal
> ./contrib/moose/scripts/update_and_rebuild_petsc.sh
> ./contrib/moose/scripts/update_and_rebuild_libmesh.sh
> rm -rf build/ install/
> make -j8
> ```
> 
> I am wondering if there is some incompatibility with the conda environment that we missed. I'm sorry this isn't working as smoothly as possible!


Recently, I've tried everything I can think of to fix this error. The results are not satisfactory.

Then, I tried the original method again. Which means building Cardinal on the MOOSE's conda environment without running 
```
./contrib/moose/scripts/update_and_rebuild_petsc.sh
./contrib/moose/scripts/update_and_rebuild_libmesh.sh
```

But this brings up new errors 
```
(moose) en@ubuntu:~/cardinal$ make
Cardinal is using HDF5 from /home/en/mambaforge3
make: /usr/bin/bash: No such file or directory
make: /usr/bin/bash: No such file or directory
make: /usr/bin/bash: No such file or directory
make: /usr/bin/bash: No such file or directory
make: /usr/bin/bash: No such file or directory
make: /usr/bin/bash: No such file or directory
make: /usr/bin/bash: No such file or directory
make: /usr/bin/bash: No such file or directory
make: /usr/bin/bash: No such file or directory
make: /usr/bin/bash: No such file or directory
make: /usr/bin/bash: No such file or directory
make: /usr/bin/bash: No such file or directory
make: /usr/bin/bash: No such file or directory
make: /usr/bin/bash: No such file or directory
make: /usr/bin/bash: No such file or directory
make: /usr/bin/bash: No such file or directory
make: /usr/bin/bash: No such file or directory
/home/en/cardinal/contrib/moose/framework/moose.mk:20: *** Build failed.  Stop.
```

**Reply by milljm** [reply]
So these two scripts successfully ran?
```pre
./contrib/moose/scripts/update_and_rebuild_petsc.sh
./contrib/moose/scripts/update_and_rebuild_libmesh.sh
```
Excellent! Can you please attache the diagnostic log which gets generated as a result of running the libmesh script? It should be located at in:
```pre
cardinal/contrib/moose/scripts/libmesh_diagnostic.log
```

This should help us understand your environment. The idea that 'bash' suddenly doesn't work is odd.

**Reply by zsh151** [reply]
1. Maybe my statement is not clear, or you misunderstood what I mean. The above error appeared when I directly use MOOSE's conda environment to build Cardinal.

I think maybe I didn't catch this error when I initially built Cardinal. And It seems more convenient to build under MOOSE's conda environment. I would prefer to fix this error. 

2. Sure, these two scripts could successfully run, even use the MOOSE's conda environment. 
```
./contrib/moose/scripts/update_and_rebuild_petsc.sh
./contrib/moose/scripts/update_and_rebuild_libmesh.sh
```

But there is another error. Below is the complete output after running `make`
[out.txt](https://github.com/neams-th-coe/cardinal/files/9309426/out.txt)

And here is the diagnostic log
[libmesh_diagnostic.log](https://github.com/neams-th-coe/cardinal/files/9309344/libmesh_diagnostic.log)

**Reply by milljm** [reply]
This is what I was trying to catch earlier:

```pre
----------------------------------- SUMMARY -----------------------------------

Package version............... : metaphysicl-1.3.3

C++ compiler.................. : mpicxx
C++ compiler flags............ : -fvisibility-inlines-hidden -fmessage-length=0 -march=nocona -mtune=haswell -ftree-vectorize -fPIC -fstack-protector-strong -fno-plt -ffunction-sections -pipe -isystem /home/en/mambaforge3/envs/moose/include -std=c++17 -Wall
Install dir................... : /home/en/mambaforge3/envs/moose/libmesh
```
The 'Install dir' is incorrect.

The install dir is pointing to the location to where you've _already_ installed the `moose-libmesh` conda package. Thus overwriting it. This is going to cause all sorts of headaches down the road.

What you need to do, is choose one:

- Install `moose-libmesh` from Conda and then _don't_ build PETSc or libMesh yourself
or
- Install `moose-mpich` from Conda and then you may build PETSc or libMesh yourself using:
     ```pre
     update_and_rebuild_petsc.sh
     update_and_rebuild_libmesh.sh
     ```

You _can_ do both, but you must be very careful with environment management. That is, create two separate Conda environments. One with perhaps only moose-mpich installed. The other with moose-libmesh. And then swap between the two when you need to.

Let us know which direction you want to take, so we can direct you properly (we need to fix what you have started before continuing).

Thanks!! And sorry you're running into all this!

**Reply by zsh151** [reply]
Thank you for your patience in answering my questions.

As what I said above, I choose to Install from Conda and then don't build PETSc or libMesh myself.
But this brings up the error
```
(moose) en@ubuntu:~/cardinal$ make
Cardinal is using HDF5 from /home/en/mambaforge3
make: /usr/bin/bash: No such file or directory
make: /usr/bin/bash: No such file or directory
make: /usr/bin/bash: No such file or directory
make: /usr/bin/bash: No such file or directory
make: /usr/bin/bash: No such file or directory
make: /usr/bin/bash: No such file or directory
make: /usr/bin/bash: No such file or directory
make: /usr/bin/bash: No such file or directory
make: /usr/bin/bash: No such file or directory
make: /usr/bin/bash: No such file or directory
make: /usr/bin/bash: No such file or directory
make: /usr/bin/bash: No such file or directory
make: /usr/bin/bash: No such file or directory
make: /usr/bin/bash: No such file or directory
make: /usr/bin/bash: No such file or directory
make: /usr/bin/bash: No such file or directory
make: /usr/bin/bash: No such file or directory
/home/en/cardinal/contrib/moose/framework/moose.mk:20: *** Build failed.  Stop.
```

Strangely, I had this error when I initially installed Cardinal. But when I build PETSc or libMesh myself (which is now proven wrong), the error goes away.

So, what should I do to fix the error above?

**Reply by LagrangeW** [reply]
> I intended to use the PETSc that ships with MOOSE's conda environment, but when I make, another error appeared. Then I tried to rebuild the PETSc. Finally, it turns out the error has nothing to do with the PETSc. I rebuild MOOSE and now that error doesn't appear,
> 
> But a new error has occurred when I run make. I have updated MOOSE but it still can't be solved.
> 
> <img alt="image" width="363" src="https://user-images.githubusercontent.com/110072634/182744571-fb18941a-3331-4bfa-96fc-2c62a4ae714e.png">

Sorry to interrupt, I've compiled cardinal using the PETSc and libMesh under MOOSE's conda environment, but encountered same problem, have you found a solution? Thanks!!

**Reply by aprilnovak** [reply]
Hi all, it seems like there may be a deeper issue with Cardinal and MOOSE's conda environment. I can think of two solutions:

- Don't use MOOSE's conda environment. We need to do more investigating to figure out what is going wrong. To be sure nothing is mixed up in your environment, please do:

```
rm -rf cardinal
conda deactivate moose

git clone https://github.com/neams-th-coe/cardinal.git
cd cardinal
./scripts/get-dependencies.sh
./contrib/moose/scripts/update_and_rebuild_petsc.sh
./contrib/moose/scripts/update_and_rebuild_libmesh.sh
```

- It's not an elegant solution, but I believe the issue is because both NekRS and MOOSE need HYPRE. If you are only interested in Cardinal to use OpenMC, you can turn the Nek part off and still use conda:

```
export ENABLE_NEK=false
make
```

**Reply by LagrangeW** [reply]
Thanks for your guidance! According to your suggestion, I didn't use MOOSE's conda environment and build PETSc and libMesh using given scripts. The process was smooth until the following error occurred in the cardinal make process:
![image](https://user-images.githubusercontent.com/75718485/185066071-da04b938-d649-4c23-80b2-6bdb0bb53908.png)
Seems something went wrong when compiling NekRS, but I really don't know what this error means, and which compiler is needed here.
Could you please give any possible solutions?
Many thanks!

**Reply by aprilnovak** [reply]
As noted on the [Getting Started](https://cardinal.cels.anl.gov/start.html) page, NekRS requires GNU compilers. We have plans very soon to relax this in the coming month to let you use clang as well. But until then, can you please show me what you see with this:

```
mpicc --version
mpicxx -version
mpif90 --version
```

**Reply by LagrangeW** [reply]
Thank you for your patience, the version information shows as below:
![image](https://user-images.githubusercontent.com/75718485/185270298-9b0aa307-dd06-49cc-83b4-6ca3dce152e0.png)
Was it because the lower version?

**Reply by aprilnovak** [reply]
No, those compilers look okay. We're aware of one issue in NekRS's compiler detection code that doesn't work on some systems. Can you please do:

```
export MPICH_FC=gfortran
cd cardinal
rm -rf build/ install/
make -j8
```

I think this should allow you to proceed.

**Reply by LagrangeW** [reply]
Exactly as you said, the executable has been built successfully by doing so! 
Thanks again for your valuable guidance!

**Reply by aprilnovak** [reply]
@zsh151 were you able to resolve the build issue?

**Reply by zsh151** [reply]
Sorry I was delayed by other things.
I have successfully built PETSc using base environment. But when I tried to build libMesh with
` ./contrib/moose/scripts/update_and_rebuild_libmesh.sh`
<img width="430" alt="image" src="https://user-images.githubusercontent.com/110072634/185786297-eb3efb1c-5671-41bf-b509-4595d4e0c86f.png">
Could you please help me out?

**Reply by AyaHegazy22** [reply]
Hi, April (the repository maintainer) is on vacation until September 15. In the meantime, could @milljm or @loganharbour perhaps assist?

**Reply by milljm** [reply]
Can you attache the libmesh_diagnostic.log? It should be located in contrib/moose/scripts/.

**Reply by zsh151** [reply]
Sorry, I'm late. It took me a long time to build petsc and libmesh. And it finally worked.

But when I ran make 
![image](https://user-images.githubusercontent.com/110072634/192144308-a0ec69b2-279b-4b60-a358-7f6fc1aaba37.png)

The full output is shown in the following file.
[output .txt](https://github.com/neams-th-coe/cardinal/files/9640775/output.txt)

Zhang

**Reply by zsh151** [reply]
Forget it, please don't mind.
There is still something wrong with libmesh.

**Reply by zsh151** [reply]
Finally, after I reinstalled libmesh and made sure there were no issues, cardinal was successfully installed.

During this process, I also encountered the same problem, i.e. `ERROR: Cannot find a supported compiler`
The above problem has also been successfully solved by running
```
export MPICH_FC=gfortran
cd cardinal
rm -rf build/ install/
make -j8
```
Thank you both for taking the trouble to answer my questions.

Zhang
